### ISSUE TITLE
Packet copies can cause undefined behavior

### ISSUE THREAD

**Issue by dberliner** [xaptum/hss#33]
The way HSS uses memcpy can cause issues when copying between different machines. This is necessary to address before upstreaming. 

From the C99 standard
>   6.5.3.2 Address and indirection operators
>    87) Among the invalid values for dereferencing a pointer by the unary * operator are a null pointer, an address inappropriately aligned for the type of object pointed to, and the address of an object after the end of its lifetime.

In practice x86 does not care nor would [our version of ARM](https://developer.arm.com/documentation/dui0472/m/compiler-command-line-options/--unaligned-access----no-unaligned-access), but other architectures like [SPARC](https://stackoverflow.com/questions/8568432/is-gccs-attribute-packed-pragma-pack-unsafe) or older ARM chips will.

Take this struct (this a bug that I forgot to update when family went from 1 to 2 bytes).
```
struct hss_payload_connect_ip {
	enum hss_family	family; /* 2 bytes */
	__u8					resvd; /* One byte */
/* Any sane compiler will place another 1 byte pad here */
	__le16					port;
/* Padding here? */
	union hss_payload_connect_ip_addr	addr;
};
```

1. Structs are not packed so the compiler can add padding. Most systems would probably pad this the same way but that's not a guarantee. Between `port` and `addr` a 64-bit system may want to align to 8 byes and insert another 4-byte pad where a 32-bit system probably won't.
2. Packing the structs is not a solution, [it is undefined behavior to de-reference a misaligned pointer](https://developer.apple.com/documentation/xcode/diagnosing_memory_thread_and_crash_issues_early/misaligned_pointer) and we would have to take special care to design structs not to be misaligned. Thats just a way to introduce problems
3. The way we transmit packets pays no attention to alignment. It occurs to me that this might be why no one else in the kernel typecasts a struct on circular buffer segment. 

The solution is to stop trying to entangle our defined format with the C structs. The members should match up in width and endianness, but memcpy needs to be used instead of typecasting and named assignments. I do not think this will have a performance hit, it's just tedious.

**Comment by dberliner** [comment]
@drbild

**Comment by dberliner** [comment]
Fixed by #35
